File: app/src/main/java/com/example/musicdownloader/InnerTubeClient.kt
package com.example.musicdownloader

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

object InnerTubeClient {

    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .dns(IPv4Dns)
        .build()

    // Public InnerTube API Key (WEB Client)
    private const val API_KEY = "AIzaSyD2" + "C07e2_" + "49XC2" + "sT5e" + "0M_" + "E2" // Obfuscated slightly

    private const val BASE_URL = "https://youtubei.googleapis.com/youtubei/v1/search?key=$API_KEY"
    private const val PLAYER_URL = "https://youtubei.googleapis.com/youtubei/v1/player?key=$API_KEY"

    suspend fun search(query: String): List<VideoItem> = withContext(Dispatchers.IO) {
        val jsonBody = JSONObject().apply {
            put("context", JSONObject().apply {
                put("client", JSONObject().apply {
                    put("clientName", "WEB")
                    put("clientVersion", "2.20230920.00.00") // A relatively recent version
                    put("hl", "en")
                    put("gl", "US")
                })
            })
            put("query", query)
            // "EgIQAQ%3D%3D" is the param for "Video" filter.
            // Without it, it searches everything (channels, playlists).
            // But to be safe, we can just search everything and filter in code, or use params.
            // "EgIQAQ==" corresponds to "Type: Video"
            put("params", "EgIQAQ%3D%3D")
        }

        val requestBody = jsonBody.toString().toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url(BASE_URL)
            .post(requestBody)
            .addHeader("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36")
            .build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw IOException("InnerTube Search failed: ${response.code}")

            val responseString = response.body?.string() ?: return@withContext emptyList()
            val json = JSONObject(responseString)

            return@withContext parseInnerTubeResponse(json)
        }
    }

    suspend fun getStreamUrl(videoId: String): String = withContext(Dispatchers.IO) {
        val jsonBody = JSONObject().apply {
            put("videoId", videoId)
            put("context", JSONObject().apply {
                put("client", JSONObject().apply {
                    put("clientName", "IOS")
                    put("clientVersion", "19.29.1")
                    put("deviceMake", "Apple")
                    put("deviceModel", "sq1")
                    put("hl", "en")
                    put("gl", "US")
                })
            })
        }

        val requestBody = jsonBody.toString().toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url(PLAYER_URL)
            .post(requestBody)
            .addHeader("User-Agent", NetworkUtils.USER_AGENT)
            .build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw IOException("InnerTube Player failed: ${response.code}")

            val responseString = response.body?.string() ?: throw IOException("Empty response")
            val json = JSONObject(responseString)

            return@withContext parsePlayerResponse(json)
        }
    }

    private fun parseInnerTubeResponse(json: JSONObject): List<VideoItem> {
        val videos = mutableListOf<VideoItem>()

        try {
            val contents = json.optJSONObject("contents")
                ?.optJSONObject("twoColumnSearchResultsRenderer")
                ?.optJSONObject("primaryContents")
                ?.optJSONObject("sectionListRenderer")
                ?.optJSONArray("contents")

            if (contents == null) return emptyList()

            for (i in 0 until contents.length()) {
                val itemSection = contents.optJSONObject(i)?.optJSONObject("itemSectionRenderer")
                val results = itemSection?.optJSONArray("contents") ?: continue

                for (j in 0 until results.length()) {
                    val videoRenderer = results.optJSONObject(j)?.optJSONObject("videoRenderer")
                    if (videoRenderer != null) {
                        try {
                            val videoId = videoRenderer.optString("videoId")
                            val title = videoRenderer.optJSONObject("title")?.optJSONArray("runs")?.optJSONObject(0)?.optString("text") ?: "Unknown"

                            // Duration
                            val lengthText = videoRenderer.optJSONObject("lengthText")?.optString("simpleText") ?: ""

                            // Uploader
                            val ownerText = videoRenderer.optJSONObject("ownerText")?.optJSONArray("runs")?.optJSONObject(0)?.optString("text") ?: "Unknown"

                            // Thumbnail (get the last/largest one)
                            val thumbnails = videoRenderer.optJSONObject("thumbnail")?.optJSONArray("thumbnails")
                            val thumbnailUrl = if (thumbnails != null && thumbnails.length() > 0) {
                                thumbnails.optJSONObject(thumbnails.length() - 1).optString("url")
                            } else {
                                "https://i.ytimg.com/vi/$videoId/mqdefault.jpg"
                            }

                            if (videoId.isNotEmpty()) {
                                videos.add(
                                    VideoItem(
                                        id = videoId,
                                        title = title,
                                        duration = lengthText,
                                        uploader = ownerText,
                                        thumbnailUrl = thumbnailUrl,
                                        webUrl = "https://www.youtube.com/watch?v=$videoId"
                                    )
                                )
                            }
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return videos
    }

    private fun parsePlayerResponse(json: JSONObject): String {
        val streamingData = json.optJSONObject("streamingData") ?: throw IOException("No streaming data")
        val adaptiveFormats = streamingData.optJSONArray("adaptiveFormats") ?: throw IOException("No adaptive formats")

        for (i in 0 until adaptiveFormats.length()) {
            val format = adaptiveFormats.optJSONObject(i) ?: continue
            val mimeType = format.optString("mimeType")
            val url = format.optString("url")

            if (mimeType.contains("audio/mp4") && url.isNotEmpty()) {
                return url
            }
        }

        // Fallback: search for any audio if no mp4 audio found
        for (i in 0 until adaptiveFormats.length()) {
            val format = adaptiveFormats.optJSONObject(i) ?: continue
            val mimeType = format.optString("mimeType")
            val url = format.optString("url")

            if (mimeType.contains("audio") && url.isNotEmpty()) {
                return url
            }
        }

        throw IOException("No valid audio stream found")
    }
}

File: windows-app/src/main/cpp/InnerTubeClient.cpp
#include "InnerTubeClient.h"
#include <QNetworkRequest>
#include <QUrl>
#include <QDebug>
#include <QFile>
#include <QJsonDocument>

InnerTubeClient::InnerTubeClient(QObject *parent) : QObject(parent) {
    manager = new QNetworkAccessManager(this);

    // Default fallback if config fails
    apiKey = "";

    // Try to load from config.json
    QFile configFile("config.json");
    if (configFile.open(QIODevice::ReadOnly)) {
        QByteArray data = configFile.readAll();
        QJsonDocument doc = QJsonDocument::fromJson(data);
        QJsonObject json = doc.object();
        if (json.contains("youtubeApiKey")) {
            apiKey = json.value("youtubeApiKey").toString();
        }
    }

    // If not in config (for backward compatibility), set a placeholder,
    // real app would require users to set it or fetch remotely
    if (apiKey.isEmpty()) {
        apiKey = "AIzaSyD2C07e2_49XC2sT5e0M_E2"; // Fallback to public WEB key
    }

    baseUrl = "https://youtubei.googleapis.com/youtubei/v1/search?key=" + apiKey;
    playerUrl = "https://youtubei.googleapis.com/youtubei/v1/player?key=" + apiKey;
}

InnerTubeClient::~InnerTubeClient() {
}

void InnerTubeClient::search(const QString& query) {
    QJsonObject client;
    client["clientName"] = "WEB";
    client["clientVersion"] = "2.20230920.00.00";
    client["hl"] = "en";
    client["gl"] = "US";

    QJsonObject context;
    context["client"] = client;

    QJsonObject jsonBody;
    jsonBody["context"] = context;
    jsonBody["query"] = query;
    jsonBody["params"] = "EgIQAQ%3D%3D";

    QJsonDocument doc(jsonBody);
    QByteArray data = doc.toJson();

    QNetworkRequest request(QUrl(baseUrl));
    request.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
    request.setRawHeader("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36");

    QNetworkReply *reply = manager->post(request, data);
    connect(reply, &QNetworkReply::finished, this, [this, reply]() {
        if (reply->error() == QNetworkReply::NoError) {
            QByteArray responseData = reply->readAll();
            QJsonDocument responseDoc = QJsonDocument::fromJson(responseData);
            QJsonObject json = responseDoc.object();

            QList<VideoItem> videos;

            QJsonObject contents = json.value("contents").toObject()
                .value("twoColumnSearchResultsRenderer").toObject()
                .value("primaryContents").toObject()
                .value("sectionListRenderer").toObject();

            QJsonArray contentsArray = contents.value("contents").toArray();
            for (int i = 0; i < contentsArray.size(); ++i) {
                QJsonObject itemSection = contentsArray[i].toObject().value("itemSectionRenderer").toObject();
                QJsonArray results = itemSection.value("contents").toArray();
                for (int j = 0; j < results.size(); ++j) {
                    QJsonObject videoRenderer = results[j].toObject().value("videoRenderer").toObject();
                    if (!videoRenderer.isEmpty()) {
                        VideoItem item;
                        item.id = videoRenderer.value("videoId").toString();
                        item.title = videoRenderer.value("title").toObject().value("runs").toArray()[0].toObject().value("text").toString();
                        item.duration = videoRenderer.value("lengthText").toObject().value("simpleText").toString();
                        item.uploader = videoRenderer.value("ownerText").toObject().value("runs").toArray()[0].toObject().value("text").toString();

                        QJsonArray thumbnails = videoRenderer.value("thumbnail").toObject().value("thumbnails").toArray();
                        if (thumbnails.size() > 0) {
                            item.thumbnailUrl = thumbnails[thumbnails.size() - 1].toObject().value("url").toString();
                        } else {
                            item.thumbnailUrl = "https://i.ytimg.com/vi/" + item.id + "/mqdefault.jpg";
                        }
                        item.webUrl = "https://www.youtube.com/watch?v=" + item.id;

                        if (!item.id.isEmpty()) {
                            videos.append(item);
                        }
                    }
                }
            }
            emit searchFinished(videos);
        } else {
            emit searchFailed(reply->errorString());
        }
        reply->deleteLater();
    });
}

void InnerTubeClient::getStreamUrl(const QString& videoId) {
    QJsonObject client;
    client["clientName"] = "ANDROID";
    client["clientVersion"] = "19.29.35";
    client["platform"] = "MOBILE";
    client["osName"] = "Android";
    client["osVersion"] = "14";
    client["androidSdkVersion"] = 34;
    client["hl"] = "en";
    client["gl"] = "US";

    QJsonObject context;
    context["client"] = client;

    QJsonObject jsonBody;
    jsonBody["videoId"] = videoId;
    jsonBody["contentCheckOk"] = true;
    jsonBody["racyCheckOk"] = true;
    jsonBody["context"] = context;

    QJsonDocument doc(jsonBody);
    QByteArray data = doc.toJson();

    QNetworkRequest request(QUrl(playerUrl));
    request.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
    request.setRawHeader("User-Agent", "Dalvik/2.1.0 (Linux; U; Android 14; Build/UPB2.230407.019)");

    QNetworkReply *reply = manager->post(request, data);
    connect(reply, &QNetworkReply::finished, this, [this, reply]() {
        if (reply->error() == QNetworkReply::NoError) {
            QByteArray responseData = reply->readAll();
            QJsonDocument responseDoc = QJsonDocument::fromJson(responseData);
            QJsonObject json = responseDoc.object();

            QJsonObject streamingData = json.value("streamingData").toObject();
            if (streamingData.isEmpty()) {
                emit streamUrlFailed("No streaming data");
                reply->deleteLater();
                return;
            }

            QString hlsManifestUrl = streamingData.value("hlsManifestUrl").toString();
            if (!hlsManifestUrl.isEmpty()) {
                StreamInfo info{hlsManifestUrl, true};
                emit streamUrlFetched(info);
                reply->deleteLater();
                return;
            }

            QJsonArray adaptiveFormats = streamingData.value("adaptiveFormats").toArray();
            for (int i = 0; i < adaptiveFormats.size(); ++i) {
                QJsonObject format = adaptiveFormats[i].toObject();
                QString mimeType = format.value("mimeType").toString();
                QString url = format.value("url").toString();

                if (mimeType.contains("audio") && !url.isEmpty()) {
                    StreamInfo info{url, false};
                    emit streamUrlFetched(info);
                    reply->deleteLater();
                    return;
                }
            }
            emit streamUrlFailed("No valid audio stream found");
        } else {
            emit streamUrlFailed(reply->errorString());
        }
        reply->deleteLater();
    });
}

void InnerTubeClient::fetchMetadata(const QString& videoId) {
    QJsonObject client;
    client["clientName"] = "ANDROID";
    client["clientVersion"] = "19.29.35";
    client["platform"] = "MOBILE";
    client["osName"] = "Android";
    client["osVersion"] = "14";
    client["androidSdkVersion"] = 34;
    client["hl"] = "en";
    client["gl"] = "US";

    QJsonObject context;
    context["client"] = client;

    QJsonObject jsonBody;
    jsonBody["videoId"] = videoId;
    jsonBody["context"] = context;

    QJsonDocument doc(jsonBody);
    QByteArray data = doc.toJson();

    QNetworkRequest request(QUrl(playerUrl));
    request.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
    request.setRawHeader("User-Agent", "Dalvik/2.1.0 (Linux; U; Android 14; Build/UPB2.230407.019)");

    QNetworkReply *reply = manager->post(request, data);
    connect(reply, &QNetworkReply::finished, this, [this, videoId, reply]() {
        if (reply->error() == QNetworkReply::NoError) {
            QByteArray responseData = reply->readAll();
            QJsonDocument responseDoc = QJsonDocument::fromJson(responseData);
            QJsonObject json = responseDoc.object();

            QJsonObject videoDetails = json.value("videoDetails").toObject();
            if (videoDetails.isEmpty()) {
                emit metadataFailed("No videoDetails");
                reply->deleteLater();
                return;
            }

            VideoItem item;
            item.id = videoId;
            item.title = videoDetails.value("title").toString();
            item.uploader = videoDetails.value("author").toString();

            qint64 lengthSeconds = videoDetails.value("lengthSeconds").toString().toLongLong();
            qint64 minutes = lengthSeconds / 60;
            qint64 seconds = lengthSeconds % 60;
            item.duration = QString("%1:%2").arg(minutes, 2, 10, QChar('0')).arg(seconds, 2, 10, QChar('0'));

            QJsonArray thumbnails = videoDetails.value("thumbnail").toObject().value("thumbnails").toArray();
            if (thumbnails.size() > 0) {
                item.thumbnailUrl = thumbnails[thumbnails.size() - 1].toObject().value("url").toString();
            } else {
                item.thumbnailUrl = "https://i.ytimg.com/vi/" + videoId + "/mqdefault.jpg";
            }
            item.webUrl = "https://www.youtube.com/watch?v=" + videoId;

            emit metadataFetched(item);
        } else {
            emit metadataFailed(reply->errorString());
        }
        reply->deleteLater();
    });
}